crates/wasi-common/src/fdentry.rs

```diff
@@ -2,40 +2,139 @@ use crate::sys::dev_null;
 use crate::sys::fdentry_impl::{
     descriptor_as_oshandle, determine_type_and_access_rights, OsHandle,
 };
+use crate::virtfs::VirtualFile;
 use crate::{wasi, Error, Result};
 use std::marker::PhantomData;
 use std::mem::ManuallyDrop;
 use std::ops::{Deref, DerefMut};
 use std::path::PathBuf;
-use std::{fs, io};
+use std::{fmt, fs, io};
+
+pub(crate) enum HandleMut<'handle> {
+    OsHandle(&'handle mut OsHandle),
+    VirtualFile(&'handle mut dyn VirtualFile),
+    Stream(OsHandleRef<'handle>),
+}
+
+impl<'descriptor> fmt::Debug for HandleMut<'descriptor> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            HandleMut::OsHandle(file) => {
+                // coerce to the target debug-printable type
+                let file: &fs::File = file;
+                write!(f, "{:?}", file)
+            }
+            HandleMut::Stream(stream) => {
+                // coerce to the target debug-printable type
+                let file: &fs::File = stream;
+                write!(f, "{:?}", file)
+            }
+            HandleMut::VirtualFile(_) => write!(f, "VirtualFile"),
+        }
+    }
+}
+
+pub(crate) enum Handle<'handle> {
+    OsHandle(&'handle OsHandle),
+    VirtualFile(&'handle dyn VirtualFile),
+    Stream(OsHandleRef<'handle>),
+}
+
+impl<'descriptor> Handle<'descriptor> {
+    pub(crate) fn try_clone(&self) -> io::Result<Descriptor> {
+        match self {
+            Handle::OsHandle(file) => file
+                .try_clone()
+                .map(|f| Descriptor::OsHandle(OsHandle::from(f))),
+            Handle::Stream(stream) => stream
+                .try_clone()
+                .map(|f| Descriptor::OsHandle(OsHandle::from(f))),
+            Handle::VirtualFile(virt) => virt.try_clone().map(|f| Descriptor::VirtualFile(f)),
```

Comment:
Just a nit really, but you can skip the closure arg here altogether:

```suggestion
            Handle::VirtualFile(virt) => virt.try_clone().map(Descriptor::VirtualFile),
```